### PR TITLE
analytics: Collect less data about controllers.

### PR DESCRIPTION
**Pull request by phire**
The name field can contain personal information, particularly in the
case of bluetooth devices on OSX which get configured with the user's
full name.

This limits the collected information to the main question we want to know
right now. The name field wasn't that helpful anyway, perhaps we can add
some better controller type detection later.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3916)

### CODE REVIEW

File: Source/Core/Core/Analytics.cpp [hunk 1]
```diff
@@ -247,12 +247,26 @@ void DolphinAnalytics::MakePerGameBuilder()
...
+	// or the official gamecube adapter.
 	builder.AddData("gcadapter-detected", GCAdapter::IsDetected());
 
-	// For privacy reasons, limit this to type of the first controller.
+	bool hasController = true;
 	// The ControllersNeedToBeCreated() check is enough to ensure GetController(0) won't return nullptr or throw exceptions.
 	if (!Pad::GetConfig()->ControllersNeedToBeCreated())
-		builder.AddData("controller-type", Pad::GetConfig()->GetController(0)->default_device.name);
+	{
+		auto& controller = Pad::GetConfig()->GetController(0)->default_device;
```
Comment (delroth) [inline Source/Core/Core/Analytics.cpp]: const

**Comment by delroth** [comment]
The type of gamepad is actually a very useful data point too. For example, to know if we should ship controller profiles, etc.

**Comment by phire** [comment]
Yes, though the name field is not particularly useful for this. Most controllers these days pretend to be an xbox360 controller on windows, which all show up with the generic name of "Gamepad"

And then on OSX it's basically freefrom. The name information coming from evdev on linux is somewhat useful, as it's generally assigned by the kernel.

File: Source/Core/Core/Analytics.cpp [hunk 2]
```diff
@@ -247,12 +247,26 @@ void DolphinAnalytics::MakePerGameBuilder()
...
+	// or the official gamecube adapter.
 	builder.AddData("gcadapter-detected", GCAdapter::IsDetected());
 
-	// For privacy reasons, limit this to type of the first controller.
+	bool hasController = true;
 	// The ControllersNeedToBeCreated() check is enough to ensure GetController(0) won't return nullptr or throw exceptions.
 	if (!Pad::GetConfig()->ControllersNeedToBeCreated())
-		builder.AddData("controller-type", Pad::GetConfig()->GetController(0)->default_device.name);
+	{
+		const auto& controller = Pad::GetConfig()->GetController(0)->default_device;
```
Comment (lioncash) [inline Source/Core/Core/Analytics.cpp]: This added code should likely be in it's own local function that returns a boolean. This sanitizing has nothing to do with the creation of the builder directly.

Comment (phire) [inline Source/Core/Core/Analytics.cpp]: Done.

File: Source/Core/InputCommon/InputConfig.cpp [hunk 3]
```diff
@@ -117,3 +117,17 @@ bool InputConfig::ControllersNeedToBeCreated() const
 {
 	return m_controllers.empty();
 }
+
+bool InputConfig::IsControllerAKeyboard(int index) const
```
Comment (lioncash) [inline Source/Core/InputCommon/InputConfig.cpp]: This name is pretty misleading considering the comment below is

`// Find anything which obviously not a controller` (which also contains a typo).

File: Source/Core/InputCommon/InputConfig.cpp [hunk 4]
```diff
@@ -117,3 +117,17 @@ bool InputConfig::ControllersNeedToBeCreated() const
...
+{
+	if (static_cast<size_t>(index) >= m_controllers.size())
+		return false;
+
+	const auto& controller = m_controllers.at(index).get()->default_device;
+
+	// Find anything which obviously not a controller
+	return ((controller.source == "Keyboard") // OSX Keyboard/Mouse
```
Comment (lioncash) [inline Source/Core/InputCommon/InputConfig.cpp]: The outer set of parentheses aren't necessary.

File: Source/Core/InputCommon/InputConfig.cpp (hunk 3 above)
Comment (phire) [inline Source/Core/InputCommon/InputConfig.cpp]: I had problems thinking of a better name that wasn't too wordy. Suggestions welcome.

Comment (RisingFog) [inline Source/Core/InputCommon/InputConfig.cpp]: How about IsNotController()?

Comment (RisingFog) [inline Source/Core/InputCommon/InputConfig.cpp]: IsNotRegularController perhaps?

Comment (JosJuice) [inline Source/Core/InputCommon/InputConfig.cpp]: That's really vague. I wouldn't understand what "regular controller" meant if I came across it.

Comment (RisingFog) [inline Source/Core/InputCommon/InputConfig.cpp]: IsNotGamepad is probably the best that it can be cut down to

File: Source/Core/InputCommon/InputConfig.cpp (hunk 4 above)
Comment (phire) [inline Source/Core/InputCommon/InputConfig.cpp]: Fixed it, they are now necessary.

File: Source/Core/InputCommon/InputConfig.h [hunk 5]
```diff
@@ -31,6 +31,7 @@ class InputConfig
 	ControllerEmu* GetController(int index);
 	void ClearControllers();
 	bool ControllersNeedToBeCreated() const;
+	bool IsControllerControlledBySomeTypeOfGamepadDevice(int index) const;
```
Comment (lioncash) [inline Source/Core/InputCommon/InputConfig.h]: `SomeTypeOf` is kind of superfluous.

File: Source/Core/InputCommon/InputConfig.cpp [hunk 6]
```diff
@@ -117,3 +117,17 @@ bool InputConfig::ControllersNeedToBeCreated() const
...
+
+bool InputConfig::IsControllerControlledBySomeTypeOfGamepadDevice(int index) const
+{
+	if (static_cast<size_t>(index) >= m_controllers.size())
+		return false;
+
+	const auto& controller = m_controllers.at(index).get()->default_device;
+
+	// Filter out anything which obviously not a gamepad
+	return !((controller.source == "Keyboard") // OSX Keyboard/Mouse
```
Comment (lioncash) [inline Source/Core/InputCommon/InputConfig.cpp]: What was particularly wrong with the more alignable and less likely to be overlooked (assuming I did the inversions correctly):

```cpp
return (controller.source != "Keyboard") &&                                     // OSX Keyboard/Mouse
       (controller.source != "XInput2")  &&                                     // Linux and BSD Keyboard/Mouse
       (controller.source != "Android" || controller.name != "Touchscreen") &&  // Android Touchscreen
       (controller.source != "DInput"  || controller.name != "Keyboard Mouse"); // Windows Keyboard/Mouse
```
?

File: Source/Core/InputCommon/InputConfig.h (hunk 5 above)
Comment (phire) [inline Source/Core/InputCommon/InputConfig.h]: `SomeTypeOfGamepadDevice` includes joysticks, steering wheels, vive controllers, wiimotes, etc.

I put it there for a reason.

File: Source/Core/InputCommon/InputConfig.cpp (hunk 6 above)
Comment (phire) [inline Source/Core/InputCommon/InputConfig.cpp]: > (assuming I did the inversions correctly):

That's always the problem. Just relax and let the compiler do the inversions for us.

File: Source/Core/InputCommon/InputConfig.h (hunk 5 above)
Comment (lioncash) [inline Source/Core/InputCommon/InputConfig.h]: `IsControllerControlledByGamepadDevice` means the same thing in that instance.

> I put it there for a reason.

Sure, but it's also unnecessary
